interceptor.go

```diff
@@ -269,57 +247,98 @@ func (i *interceptor) WrapStreamingHandler(next connect.StreamingHandlerFunc) co
+		)
+		// lastly, inject the context span to the carrier
+		i.config.propagator.Inject(ctx, carrier)
+		defer span.End()
 		streamingHandler := &streamingHandlerInterceptor{
 			StreamingHandlerConn: conn,
 			receive: func(msg any, conn connect.StreamingHandlerConn) error {
-				return state.receive(ctx, instr, msg, conn)
+				size, unlock, err := state.receive(msg, conn, protocol)
+				defer unlock()
+				if errors.Is(err, io.EOF) {
+					return err
+				}
+				span.AddEvent(messageKey,
+					trace.WithAttributes(attributeFilter(req,
+						semconv.MessageTypeReceived,
+						semconv.MessageUncompressedSizeKey.Int(size),
+						semconv.MessageIDKey.Int(state.receivedCounter),
+					)...,
+					),
+				)
+				// In WrapStreamingHandler the 'receive' is a request message.
+				state.attributes = attributeFilter(req, state.attributes...)
+				instrumentation.requestSize.Record(ctx, int64(size), state.attributes...)
+				instrumentation.requestsPerRPC.Record(ctx, 1, state.attributes...)
+				return err
 			},
 			send: func(msg any, conn connect.StreamingHandlerConn) error {
-				return state.send(ctx, instr, msg, conn)
+				size, unlock, err := state.send(msg, conn, protocol)
+				defer unlock()
+				if errors.Is(err, io.EOF) {
+					return err
+				}
+				span.AddEvent(messageKey,
+					trace.WithAttributes(attributeFilter(req,
+						semconv.MessageTypeSent,
+						semconv.MessageUncompressedSizeKey.Int(size),
+						semconv.MessageIDKey.Int(state.sentCounter),
+					)...,
+					),
+				)
+				// In WrapStreamingHandler the 'send' is a response message.
+				state.attributes = attributeFilter(req, state.attributes...)
+				instrumentation.responsesPerRPC.Record(ctx, 1, state.attributes...)
+				instrumentation.responseSize.Record(ctx, int64(size), state.attributes...)
+				return err
 			},
 		}
 		err = next(ctx, streamingHandler)
-		state.attrs = append(state.attrs, statusCodeAttribute(state.protocol, err))
-		instr.duration.Record(ctx, i.config.now().Sub(requestStartTime).Milliseconds(), state.attrs...)
+		state.attributes = append(state.attributes, statusCodeAttribute(protocol, err))
+		state.attributes = attributeFilter(req, state.attributes...)
+		span.SetAttributes(state.attributes...)
+		span.SetStatus(spanStatus(err))
+		instrumentation.duration.Record(ctx, i.config.now().Sub(requestStartTime).Milliseconds(), state.attributes...)
 		return err
 	}
 }
 
-func parseAddress(address string) []attribute.KeyValue {
-	if addrPort, err := netip.ParseAddrPort(address); err == nil {
-		return []attribute.KeyValue{
-			semconv.NetPeerIPKey.String(addrPort.Addr().String()),
-			semconv.NetPeerPortKey.Int(int(addrPort.Port())),
-		}
-	}
-	if host, port, err := net.SplitHostPort(address); err == nil {
-		portint, err := strconv.Atoi(port)
-		if err != nil {
-			return []attribute.KeyValue{
-				semconv.NetPeerNameKey.String(host),
-				semconv.NetPeerPortKey.Int(portint),
-			}
-		}
+func protocolToSemConv(peer string) string {
```

Comment:
nit: can we rename this param to `protocol`?

Reply:
6069b13